### PR TITLE
DBZ-2784 Support DATE column types with Oracle Logminer

### PR DESCRIPTION
**Pull request by Naros**
https://issues.redhat.com/browse/DBZ-2784

This PR is based on the work in PR #226, so it should be merged only after that has been merged.

### CODE REVIEW

**Comment by gunnarmorling** [comment]
This one would need a rebase, too, @Naros!

**Comment by Naros** [comment]
@gunnarmorling Rebased & ready for review.

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/OracleChangeRecordValueConverter.java [hunk 1]
```diff
@@ -233,14 +233,32 @@ private Object convertToLocalDateTime(Column column, Field fieldDefn, Object val
...
                     dateTime = LocalDateTime.from(TIMESTAMP_FORMATTER.parse(dateText.trim()));
                 }
-                // todo: DBZ-137 this originally converted to ZoneId.systemDefault() but this should be GMT
-                // this is so that this behavior is compatible with the Xstreams implementation
-                return dateTime.atZone(ZoneId.of("GMT")).toInstant().toEpochMilli() * 1000; // timestamp(6) is converted as microTimestamp
+                return getDateTimeWithPrecision(column, dateTime);
+            }
+            else if (valueString.toLowerCase().startsWith("to_date")) {
+                dateText = valueString.substring("to_date".length() + 1, valueString.length() - 1);
+                String[] parts = dateText.split(",");
```
Comment (gunnarmorling) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/OracleChangeRecordValueConverter.java:240]: Given this is on the hot path, it's worth extracting a pre-compiled `Pattern` which is kept as a static field.

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/OracleChangeRecordValueConverter.java [hunk 2]
```diff
@@ -233,14 +233,32 @@ private Object convertToLocalDateTime(Column column, Field fieldDefn, Object val
...
+                return getDateTimeWithPrecision(column, dateTime);
             }
         }
         return value;
     }
 
+    private Object getDateTimeWithPrecision(Column column, LocalDateTime dateTime) {
+        if (adaptiveTimePrecisionMode || adaptiveTimeMicrosecondsPrecisionMode) {
+            if (getTimePrecision(column) <= 3) {
+                return dateTime.atZone(ZoneId.of("GMT")).toInstant().toEpochMilli();
```
Comment (gunnarmorling) [inline debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/OracleChangeRecordValueConverter.java:252]: Let's create a constant for the GMT zone id.

File: debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/OracleConnectorIT.java [hunk 3]
```diff
@@ -1048,11 +1048,64 @@ public void shouldAvroSerializeColumnsWithSpecialCharacters() throws Exception {
         final SourceRecord stream = streams.recordsForTopic("server1.DEBEZIUM.COLUMNS_TEST").get(0);
         VerifyRecord.isValidInsert(stream, "ID", 2);
 
-        after = ((Struct) stream.value()).getStruct(Envelope.FieldName.AFTER);
+        after = ((Struct) stream.value()).getStruct(AFTER);
         assertThat(after.getInt32("ID")).isEqualTo(2);
         assertThat(after.get("AMOUNT_")).isEqualTo(VariableScaleDecimal.fromLogical(after.schema().field("AMOUNT_").schema(), BigDecimal.valueOf(23456.78d)));
     }
 
+    @Test
```
Comment (gunnarmorling) [inline debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/OracleConnectorIT.java:1056]: `@TestFor` missing.

File: debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/OracleConnectorIT.java [hunk 4]
```diff
@@ -1048,11 +1048,64 @@ public void shouldAvroSerializeColumnsWithSpecialCharacters() throws Exception {
...
+            assertThat(snapshotAfter.get("ORDER_DATE")).isEqualTo(1519257600000L);
+
+            waitForStreamingRunning(TestHelper.CONNECTOR_NAME, TestHelper.SERVER_NAME);
+
+            connection.execute("INSERT INTO debezium.orders VALUES (10, '22-FEB-2018')");
+            connection.execute("COMMIT");
+
+            final SourceRecords streamRecords = consumeRecordsByTopic(1);
+            final List<SourceRecord> orders = streamRecords.recordsForTopic("server1.DEBEZIUM.ORDERS");
+            assertThat(orders.size()).isEqualTo(1);
```
Comment (gunnarmorling) [inline debezium-connector-oracle/src/test/java/io/debezium/connector/oracle/OracleConnectorIT.java:1098]: Nit: assertThat(orders).hasSize(1) is a bit more idiomatic.

**Review by gunnarmorling — APPROVED** [review]
@Naros some last small comments, otherwise good to go. Thanks!

File: debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/OracleChangeRecordValueConverter.java (hunk 1 above)
Comment (Naros) [reply debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/OracleChangeRecordValueConverter.java:240]: I assume you're referring specifically to the else-if check?

**Comment by Naros** [comment]
Going to rebase to resolve conflicts & then this is ready.

**Comment by gunnarmorling** [comment]
Applied; thx!

**Pull request merged**